dashboard/src/components/App.js

```diff
@@ -0,0 +1,205 @@
+            console.log('Unknown message key: ' + key);
+        }
+      } catch (error) {
+        console.log('Unable to handle message:\n' + JSON.stringify(message));
+        console.error(error);
+      }
+    });
+  }
+
+
+  // DEVICE + DATA INITIALIZATION
+  function processDeviceInitialization(device) {
+    if (!connectedDevices.some(connectedDevice => connectedDevice.id === device.id)) {
+      console.log('Device initialization received: ' + JSON.stringify(device));
+      onDeviceWithNewIdConnected(device);
+    }
+  }
+
+  function processDataRecordingContainer(partialDataRecordingContainer) {
+    if (typeof selectedDevice === 'undefined' || partialDataRecordingContainer.deviceInfo.id !== selectedDevice.id) {
+      // console.log('Not processing data recording container from: ' + partialDataRecordingContainer.deviceInfo.id);
+      return;
+    }
+
+    let delay = Date.now() - partialDataRecordingContainer.endTimestamp;
+    if (Math.abs(timestampOffset - delay) > DELAY_LIMIT) {
+      timestampOffset = delay - DELAY_LIMIT;
+      console.log("Updated timestamp offset to " + timestampOffset);
+    }
+
+    // updateStatusText('Processing partial data recording with ' + delay + 'ms delay')
+
+    partialDataRecordingContainer.recordings.forEach(function (dataRecording) {
+      if (dataRecordingContainer.getData(dataRecording.dataId).length === 0) {
+        onDataWithNewIdReceived(dataRecording.dataId);
+      }
+      dataRecordingContainer.addDataRecording(dataRecording);
+    });
+    dataRecordingContainer.trim();
+  }
+
+  function onDeviceWithNewIdConnected(device) {
+
+    // update the connected devices array, place the new device first
+    updateConnectedDeviceList((oldConnectedDeviceList) => {
+      /* var connectionStatus = true;
+
+      oldConnectedDeviceList.forEach(function (oldDevice) {
+          if (oldDevice.id === device.id) {
+            connectionStatus = false;
+          }
+      }); */
+
+      const connectionStatus = !oldConnectedDeviceList.some(oldDevice => oldDevice.id === device.id);
+
+      if (connectionStatus) {
+        M.toast({
+          html: device.name + ' connected'
+        });
+        return oldConnectedDeviceList.concat(device);
+      } else {
+        return oldConnectedDeviceList;
+      }
+    });
+  }
+
+  function onDataWithNewIdReceived(id) {
+    // first recording of data with that ID
+    console.log('Received first recording of data with ID: ' + id);
+
+    let ids = dataRecordingContainer.getIds().sort();
+
+    updateDataList([]);
+
+    // append available IDs as options
+    ids.forEach(id => {
+      let optionText = DataRecordingContainer.getReadableId(id);
+
+      updateDataList((oldDataList) => {
+        // return oldDataList.concat({id: optionValue, optionText: optionText});
+        return [...oldDataList, {
+          id,
+          optionText
+        }];
+      });
+    });
+  }
+
+
+  // HANDLE SELECT-CHANGE
+  function handleDeviceChange(event) {
```

Comment:
functions passed as props to components (callbacks) should be wrapped in useCallback hook to optimize for performance.

this applies to:
- handleDeviceChange
- handleDataChange